# ~~First~~ Fast Python Notebook

<div style="max-width: 640px">

An accelerated guide to analyzing data with the <a href="https://www.python.org/">Python</a> programming language and a <a href="https://jupyter.org/">Jupyter</a> notebook

</div>
    
By [Ben Welsh](https://palewi.re/who-is-ben-welsh/)

<div style="max-width: 640px">

First developed in 2016, <a href="https://palewi.re/docs/first-python-notebook/">"First Python Notebook"</a> is a tutorial that guides students through a data-driven investigation of money in California politics. It is most commonly taught as a six-hour, in-person class. This document is an abbreviated spinoff intended to be taught online in two to three hours.

You will learn just enough of the Python computer programming language to work with the <a href="https://pandas.pydata.org/">pandas</a> library, a popular open-source tool for analyzing data. The course will teach you how to read, filter, join, group, aggregate and rank structured data by recreating a helicopter accident analysis <a href="https://github.com/datadesk/helicopter-accident-analysis">published by the Los Angeles Times</a>.
    
</div>

## What is a Jupyter notebook?

<div style="max-width: 640px">

<p><img src="https://palewi.re/docs/first-python-notebook/_static/img/labpreview.webp" alt="jupyter"></p>
<p>A <a href="https://jupyter.org/">Jupyter</a> notebook is a browser-based interface where you can write, run, remix and republish code. It is free software that anyone can install and run.</p>
<p><a href="https://nbviewer.jupyter.org/github/robertodealmeida/notebooks/blob/master/earth_day_data_challenge/Analyzing%20whale%20tracks.ipynb">Scientists</a>, <a href="https://nbviewer.jupyter.org/github/nealcaren/workshop_2014/blob/master/notebooks/5_Times_API.ipynb">scholars</a>, <a href="https://github.com/rsvp/fecon235/blob/master/nb/fred-debt-pop.ipynb">investors</a> and <a href="https://netflixtechblog.com/notebook-innovation-591ee3221233">corporations</a> use Jupyter to create and share their research. It is also used by journalists to develop stories and show their work. Examples include:</p>
<ul>
<li><a href="https://github.com/BuzzFeedNews/2016-01-tennis-betting-analysis/blob/master/notebooks/tennis-analysis.ipynb">“The Tennis Racket”</a> by BuzzFeed and the BBC</li>
<li><a href="https://github.com/propublica/compas-analysis/blob/master/Compas%20Analysis.ipynb">“Machine bias”</a> by ProPublica</li>
<li><a href="https://github.com/asuozzo/arcos-opioid-analysis-vt">“As Opioid Crisis Ramped Up, Pills Flowed Into Vermont by the Millions”</a> by Seven Days</li>
<li><a href="https://github.com/datadesk/notebooks">More than 35 different notebooks</a> published by the Los Angeles Times</li>
</ul>
<p>There are numerous ways to install and configure Jupyter notebooks. This class is taught using <a href="https://jupyterlite.readthedocs.io/">JupyterLite</a> a lightweight distribution that runs entirely in your web browser. For instructions on how to install a more powerful version on your computer consult <a href="https://palewi.re/docs/first-python-notebook/jupyter_desktop.html">the full edition of &quot;First Python Notebook.&#39;</a></p>
<p>Once you have this notebook up and running, you&#39;re ready to write Python in a code cell. Do not stress. There is nothing too fancy about it. You can start by just doing a little simple math.</p>
<p>Select the box below, then hit the play button in the toolbar above the notebook or hit <code>SHIFT+ENTER</code> on your keyboard.</p>
    
</div>

In [1]:
2+2

4

<div style="max-width:630px">

There. You have just run your first Python code. You have entered two integers and added them together using the plus sign operator.

Not so bad, right? Now try writing in your own math problem in the next cell. Maybe `2+3` or `2+200`. Whatever strikes your fancy. After you've typed it in, hit the play button or `SHIFT+ENTER`.

</div>

<div style="max-width: 640px;">

This to-and-fro of writing Python code in a cell and then running it is the rhythm of working in a notebook. If you get an error after you run a cell, look carefully at your code and see that it exactly matches what’s been written in the example.
    
Here's an example of a error that I've added intentionally:
    
</div>

In [4]:
2+2+

SyntaxError: invalid syntax (4150814810.py, line 1)

<div style="max-width: 640px;">

Don’t worry. Code crashes are a normal part of life for computer programmers. They’re usually caused by small typos that can be quickly corrected.
    
</div>

In [3]:
2+2+2

6

<div style="max-width: 640px;">

<p>Over time you will gradually stack cells to organize an analysis that runs from top to bottom. The cells can contain variables, functions and other Python tools.</p>
<div class="alert alert-block alert-warning">
<p>If you’ve never written code before, we recommend <a href="https://docs.python.org/3/tutorial/introduction.html">&quot;An Informal Introduction to Python&quot;</a> and subsequent sections of python.org’s tutorial.</p>
</div>
<p>A simple example would be storing your number in a variable in one cell:</p>

    
</div>

In [4]:
number = 2

Then adding it to another number in the next:

In [5]:
number + 3

5

Change the `number` value to 3 and run both cells again. Instead of 5, it should now output 6.

In [5]:
number = 3

In [6]:
number + 3

6

Now try defining your own numeric variable and doing some math with it. You can name it whatever you want. Want to try some other math operations? The `-` sign does subtraction. Multipication is `*`. Division is `/`.

Once you’ve got the hang of making the notebook run, you’re ready to introduce pandas, a powerful Python analysis library that can do a whole lot more than add a few numbers together.

## What is pandas? 

<div style="max-width: 640px">
  
<p><img src="https://palewi.re/docs/first-python-notebook/_static/img/pandas-pypi.png" alt="pandas on the Python Package Index"></p>
<p>Lucky for us, Python is filled with functions to do almost anything you’d want to do with a programming language: <a href="http://docs.python-requests.org/">navigate the web</a>, <a href="https://docs.python.org/2/library/csv.html">parse data</a>, <a href="http://www.sqlalchemy.org/">interact with a database</a>, <a href="https://www.scipy.org/">run fancy statistics</a>, <a href="https://www.djangoproject.com/">build a pretty website</a> and <a href="https://www.crummy.com/software/BeautifulSoup/">so</a> <a href="http://www.nltk.org/">much</a> <a href="https://pillow.readthedocs.io/en/stable/">more</a>.</p>
<p>Creative people have put these tools to work to get a <a href="https://www.python.org/about/success/">wide range of things</a> done in the academy, the laboratory and even in outer space.</p>
<p>Some of those tools are included in a toolbox that comes with the language, known as the standard library. Others have been built by members of Python’s developer community and need to be separately downloaded and installed. One third-party tool that’s important for this class is called <a href="https://pandas.pydata.org/">pandas</a>. Invented by programmers at a <a href="https://www.aqr.com/">financial investment firm</a>, it has become a leading open-source library for accessing and analyzing data.</p>
<p>Here’s how to use pandas yourself. Run the following:</p>
    
</div>

In [8]:
import pandas

<div style="max-width: 640px">

<p>If nothing happens, that’s good. It means you have it installed and ready as to use.</p>
<div class="alert alert-block alert-warning">
<p>Since pandas is created by a third party independent from the core Python developers, it may not be available by default if you manually installed Python and Jupyter. It’s available here because JupyterLite, whose developers have curated a list of common utilities to include with their distribution. Consult our <a href="https://palewi.re/docs/first-python-notebook/appendix/index.html">advanced installation guide</a> if the cell above threw an error.</p>
</div>
<p>Now let&#39;s run the same code again, but with a small addition.</p>

    
</div>

In [10]:
import pandas as pd

<div style="max-width: 640px">

This will alias the pandas library at the shorter variable name of `pd`. This is standard practice in the pandas community. You will frequently see examples of pandas code online using pd as shorthand. It’s not required, but it’s good to get in the habit so that your code will be better understood by other computer programmers.

Those two little letters contain dozens of data analysis tools that we’ll use in future lessons. They can import massive data files, compute advanced statistics, filter, sort, rank and do just about anything else you’d want to do.

We’ll get to all of that soon enough, but let’s start out with something simple. Let's run some simple stats.

## Calculating descriptive statistics

Start by making a list of numbers in a new notebook cell. To keep things simple, we'll start with all of the even numbers between zero and ten. Note the variable name I've assigned. Then press play.
    
</div>

In [15]:
my_list = [2, 4, 6, 8]

<div style="max-width: 640px">

If you’re a skilled Python programmer, you can do some cool stuff with any list, including run statistics. But if you hand over to pandas instead, you’ll be impressed by how easily you can analyze the data without much computer code.


In this case, it’s as simple as converting that plain Python list into what pandas calls a [`Series`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html). Here’s how to make it happen:
    
</div>

In [16]:
my_series = pd.Series(my_list)

<div style="max-width: 640px">

Once the data becomes a `Series`, you can immediately run a wide range of <a href="https://en.wikipedia.org/wiki/Descriptive_statistics">descriptive statistics</a>. Let’s try a few.

First, let’s sum all the numbers.
    
</div>

In [17]:
my_series.sum()

20

Then find the maximum value.

In [18]:
my_series.max()

8

The minimum value.

In [19]:
my_series.min()

2

How about the average, which also known as the mean?

In [20]:
my_series.mean()

5.0

The median?

In [21]:
my_series.median()

5.0

The standard deviation?

In [22]:
my_series.std()

2.581988897471611

Finally, all of the above, plus a little more about the distribution, in one simple command.

In [18]:
my_series.describe()

count    4.000000
mean     5.000000
std      2.581989
min      2.000000
25%      3.500000
50%      5.000000
75%      6.500000
max      8.000000
dtype: float64

<div style="max-width: 640px">

Before you move on, go back the `my_list` variable and change the list. Maybe add a few more values. Or switch to odds. Then rerun all the cells above. You'll see all the statistics update to reflect the different dataset.

Substitute in a series of 10 million records and your notebook would calculate all the same statistics without you needing to write any more code. Once your data, however large or complex, is imported into pandas, simple statistics become a snap.

</div>

<div style="max-width:640px">

## Introducing DataFrames
    
Now it’s time to get our hands on some real data. In 2018, the Los Angeles Times published an investigation headlined, <a href="https://www.latimes.com/projects/la-me-robinson-helicopters/">"The Robinson R44, the world’s best-selling civilian helicopter, has a long history of deadly crashes".</a>
    
It reported that the Robinson R44 led all major models with the highest fatal accident rate from 2006 to 2016. The analysis was <a href="https://github.com/datadesk/helicopter-accident-analysis">published on GitHub</a> as a series of Jupyter notebooks. 

The analysis was based on two key datasets: 
   
1. The National Transportation Safety Board's <a href="https://www.ntsb.gov/_layouts/ntsb.aviation/index.aspx">Aviation Accident Database</a>
2. The Federal Aviation Administration's <a href="https://www.faa.gov/data_research/aviation_data_statistics/general_aviation/">General Aviation and Part 135 Activity Survey</a>

After a significant amount of work gathering and cleaning the source data, the number of accidents for each helicopter model were normalized using the flight hours estimates in the survey. For the purposes of this demonstration, we will read in tidied versions of each file that are ready for analysis.
    
The data are structured in rows of comma-separated values. This is known as a [CSV file](https://en.wikipedia.org/wiki/Comma-separated_values). It is the most common way you will find data published online.

The pandas library is able to read in files from a variety formats, including CSV. In our next cell, we'll use pandas' `read_csv` method to read in `ntsb-accidents.csv`.
    
</div>

In [27]:
# df = pd.read_csv("raw-accidents.csv")
# survey = pd.read_csv("faa-survey.csv")
# years = [2006, 2007, 2008, 2009, 2010, 2012, 2013, 2014, 2015, 2016]
# df = df[df.in_usa == True]
# df = df[
#     df.year.isin(years) &
#     df.latimes_make_and_model.isin(survey.latimes_make_and_model.unique())
# ]
# df.to_csv("ntsb-accidents.csv", index=False)

In [29]:
pd.read_csv("ntsb-accidents.csv")

event_id  aircraft_id                    ntsb_make ntsb_model  \
0    20121010X41638            1                         BELL       206B   
1    20130222X12624            1                   EUROCOPTER  AS 350 B2   
2    20140127X44720            1                         BELL     206L 3   
3    20091012X15520            1    MCDONNELL DOUGLAS HELI CO      369FF   
4    20161116X00108            1                       HUGHES  369D/500D   
..              ...          ...                          ...        ...   
158  20070802X01089            1                   EUROCOPTER   AS350-B2   
159  20070802X01089            2                   EUROCOPTER   AS350-B2   
160  20060605X00682            1                         BELL     206L-3   
161  20140717X70001            1                   AGUSTA SPA      A109E   
162  20160612X85856            1  ROBINSON HELICOPTER COMPANY     R44 II   

    ntsb_number  year               date             city state country  \
0    CEN13FA009  2012  10/10/12 00:00:00          CROWLEY    LA     USA   
1    CEN13FA174  2013  02/22/13 00:00:00    OKLAHOMA CITY    OK     USA   
2    CEN14FA122  2014  01/27/14 00:00:00             SILT    CO     USA   
3    WPR10FA012  2009  10/12/09 00:00:00           FOSSIL    OR     USA   
4    WPR17FA021  2016  11/15/16 00:00:00            PUKOO    HI     USA   
..          ...   ...                ...              ...   ...     ...   
158  LAX07MA231  2007  07/27/07 00:00:00          PHOENIX    AZ     USA   
159  LAX07MA231  2007  07/27/07 00:00:00          PHOENIX    AZ     USA   
160  DFW06IA145  2006  05/30/06 00:00:00  GRAND ISLE 43AA    GM     USA   
161  CEN14FA369  2014  07/17/14 00:00:00          NEWKIRK    NM     USA   
162  CEN16FA215  2016  06/12/16 00:00:00        JONESBORO    AR     USA   

     total_fatalities  in_usa  pilot_total       latimes_make latimes_model  \
0                 1.0    True          1.0               BELL           206   
1                 2.0    True          1.0             AIRBUS           350   
2                 3.0    True          1.0               BELL           206   
3                 1.0    True          1.0  MCDONNELL DOUGLAS           369   
4                 2.0    True          1.0             HUGHES           369   
..                ...     ...          ...                ...           ...   
158               4.0    True          1.0             AIRBUS           350   
159               4.0    True          1.0             AIRBUS           350   
160               1.0    True          1.0               BELL           206   
161               3.0    True          1.0             AGUSTA           109   
162               1.0    True          1.0           ROBINSON           R44   

     count latimes_make_and_model  
0    112.0               BELL 206  
1      8.0             AIRBUS 350  
2      3.0               BELL 206  
3      3.0  MCDONNELL DOUGLAS 369  
4      1.0             HUGHES 369  
..     ...                    ...  
158    3.0             AIRBUS 350  
159    3.0             AIRBUS 350  
160   20.0               BELL 206  
161    2.0             AGUSTA 109  
162   22.0           ROBINSON R44  

[163 rows x 17 columns]

<div style="max-width:640px">

You should see a big table like the one above. It is a DataFrame where pandas has structured the CSV data into rows and columns, just like Excel or other spreadsheet software might.

A major advantage of Jupyter over spreadsheets is that rather than manipulating the data through a haphazard series of clicks and keypunches we will be gradually grinding it down using a computer programming script that is transparent and reproducible.

In order to do more with your DataFrame, we need to store it so it can be reused in subsequent cells. We can do this by saving in a variable, which is a fancy computer programming word for a named shortcut where we save our work as we go.

</div>

In [30]:
accident_list = pd.read_csv("ntsb-accidents.csv")

<div style="max-width:640px">

After you run it, you shouldn’t see anything. That’s a good thing. It means our DataFrame has been saved under the name `accident_list`, which we can now begin interacting with in the cells that follow.

We can do this by calling “methods” that pandas makes available to all DataFrames. You may not have known it at the time, but `read_csv` is one of these methods. There are dozens more that can do all sorts of interesting things. Let’s start with some easy ones that analysts use all the time.

To preview the first few rows of the dataset, try the <a href="https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.head.html">`head`</a> method.
    
</div>

In [31]:
accident_list.head()

event_id  aircraft_id                  ntsb_make ntsb_model  \
0  20121010X41638            1                       BELL       206B   
1  20130222X12624            1                 EUROCOPTER  AS 350 B2   
2  20140127X44720            1                       BELL     206L 3   
3  20091012X15520            1  MCDONNELL DOUGLAS HELI CO      369FF   
4  20161116X00108            1                     HUGHES  369D/500D   

  ntsb_number  year               date           city state country  \
0  CEN13FA009  2012  10/10/12 00:00:00        CROWLEY    LA     USA   
1  CEN13FA174  2013  02/22/13 00:00:00  OKLAHOMA CITY    OK     USA   
2  CEN14FA122  2014  01/27/14 00:00:00           SILT    CO     USA   
3  WPR10FA012  2009  10/12/09 00:00:00         FOSSIL    OR     USA   
4  WPR17FA021  2016  11/15/16 00:00:00          PUKOO    HI     USA   

   total_fatalities  in_usa  pilot_total       latimes_make latimes_model  \
0               1.0    True          1.0               BELL           206   
1               2.0    True          1.0             AIRBUS           350   
2               3.0    True          1.0               BELL           206   
3               1.0    True          1.0  MCDONNELL DOUGLAS           369   
4               2.0    True          1.0             HUGHES           369   

   count latimes_make_and_model  
0  112.0               BELL 206  
1    8.0             AIRBUS 350  
2    3.0               BELL 206  
3    3.0  MCDONNELL DOUGLAS 369  
4    1.0             HUGHES 369

It does the first five by default. If you want a different number, submit it as an input.

In [32]:
accident_list.head(1)

event_id  aircraft_id ntsb_make ntsb_model ntsb_number  year  \
0  20121010X41638            1      BELL       206B  CEN13FA009  2012   

                date     city state country  total_fatalities  in_usa  \
0  10/10/12 00:00:00  CROWLEY    LA     USA               1.0    True   

   pilot_total latimes_make latimes_model  count latimes_make_and_model  
0          1.0         BELL           206  112.0               BELL 206

To get a look at all of the columns and what type of data they store, try the [`info`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.info.html) method.

In [33]:
accident_list.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 163 entries, 0 to 162
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   event_id                163 non-null    object 
 1   aircraft_id             163 non-null    int64  
 2   ntsb_make               163 non-null    object 
 3   ntsb_model              163 non-null    object 
 4   ntsb_number             163 non-null    object 
 5   year                    163 non-null    int64  
 6   date                    163 non-null    object 
 7   city                    163 non-null    object 
 8   state                   162 non-null    object 
 9   country                 163 non-null    object 
 10  total_fatalities        163 non-null    float64
 11  in_usa                  163 non-null    bool   
 12  pilot_total             163 non-null    float64
 13  latimes_make            163 non-null    object 
 14  latimes_model           163 non-null    ob

Look carefully at the results and you'll see we have 163 fatal accidents.

## Inspect columns

<div style="max-width: 640px;">

To see the contents of a column separate from the rest of the DataFrame, add the column’s name to the DataFrame’s variable following a period. We’ll begin with the `latimes_make_and_model` column, which records the standardized name of the helicopter that crashed.

</div>

In [36]:
accident_list.latimes_make_and_model

0                   BELL 206
1                 AIRBUS 350
2                   BELL 206
3      MCDONNELL DOUGLAS 369
4                 HUGHES 369
               ...          
158               AIRBUS 350
159               AIRBUS 350
160                 BELL 206
161               AGUSTA 109
162             ROBINSON R44
Name: latimes_make_and_model, Length: 163, dtype: object

<div style="max-width: 640px;">

That will list the column out as a `Series`, just like the ones we created from scratch earlier. Just as we did then, you can now start tacking on additional methods that will analyze the contents of the column.


<div class="alert alert-block alert-warning">    
    <p>You can also access columns a second way, like this: accident_list['latimes_make_and_model'].</p><p>This method isn’t as pretty, but it’s required if your column has a space in its name, which would break the simpler dot-based method.</p>
</div>
    
In this case, the column is filled with characters. So we don’t want to calculate statistics like the median and average, as we did before.

There’s another built-in pandas tool that will total up the frequency of values in a column. The method is called `value_counts` and it’s just as easy to use as sum, min or max. All you need to do it is add a period after the column name and chain it on the tail end of your cell.

Run the code and you should see the locations ranked by their number of sites.

</div>

In [37]:
accident_list.latimes_make_and_model.value_counts()

ROBINSON R44             38
BELL 206                 30
AIRBUS 350               29
ROBINSON R22             20
HUGHES 369               13
BELL 407                 13
MCDONNELL DOUGLAS 369     6
SCHWEIZER 269             5
AIRBUS 135                4
SIKORSKY 76               2
AGUSTA 109                2
AIRBUS 130                1
Name: latimes_make_and_model, dtype: int64

<div style="max-width: 640px;">

Congratulations, you've made your first finding. With that little line of code, you've calculated an important fact: During the period being studied, the Robinson R44 had more fatal accidents than any other helicopter.
    
You may notice that even though the result has two columns, pandas did not return a clean-looking table in the same way as `head` did for our DataFrame. That’s because our column, a Series, acts a little bit different than the DataFrame created by `read_csv`.

In most instances, if you have an ugly Series generated by a method like `value_counts` and you want to convert it into a pretty DataFram,e you can do so by tacking on the `reset_index` method on the end.

</div>

In [39]:
accident_list.latimes_make_and_model.value_counts().reset_index()

index  latimes_make_and_model
0            ROBINSON R44                      38
1                BELL 206                      30
2              AIRBUS 350                      29
3            ROBINSON R22                      20
4              HUGHES 369                      13
5                BELL 407                      13
6   MCDONNELL DOUGLAS 369                       6
7           SCHWEIZER 269                       5
8              AIRBUS 135                       4
9             SIKORSKY 76                       2
10             AGUSTA 109                       2
11             AIRBUS 130                       1

<div style="max-width: 640px;">

Why does `Series` behave differently than a dataframe? Why does `reset_index` have such a weird name?

Like so much in computer programming, the answer is simply, “because the people who created the library said so.” It’s important to learn that all open-source programming tools are made by humans, and humans have their quirks. Over time you’ll see pandas has more than a few.

As a beginner, you should just accept the oddities and keep moving. As you get more advanced, if there’s something about the system you think could be improved you should consider <a href="https://pandas.pydata.org/pandas-docs/stable/development/contributing.html">contributing</a> to the Python code that operates the library.
    
Before we move on to the next chapter, here's a challenge. See if you can answer a few more questions a journalist might ask about our dataset. All four of the questions below can be answered using only tricks we've covered thus far. See if you can do it.
    
</div>

1. What was the total number of fatalities?

2. Which helicopter maker had the most accidents?

3. What was the total number of helicopter accidents by year?

4. What state had the most helicopter accidents?

## Filtering down the dataset

<div style="max-width: 640px;">

The most common way to filter a DataFrame is to pass an expression as an “index” that can be used to decide which records should be kept and which discarded. You write the expression by combining a column on your DataFrame with an <a href="https://en.wikipedia.org/wiki/Operator_(computer_programming)">“operator”</a> like == or > or < and a value to compare against each row.

<div class="alert alert-block alert-warning">
    <p>If you are familiar with writing <a href="https://en.wikipedia.org/wiki/SQL">SQL</a> to manipulate databases, pandas’ filtering system is somewhat similar to a WHERE query. The <a href="https://pandas.pydata.org/pandas-docs/stable/getting_started/comparison/comparison_with_sql.html#where">official pandas documentation</a> offers direct translations between the two.</p>
</div>
    
Let's try filtering against the `state` field. Save one of the values listed above into a variable. This will allow us to reuse it later.

</div>

In [41]:
my_state = "IA"

<div style="max-width: 640px;">

In the next cell we will ask pandas to narrow down our list of sites to just those that list the location we’re interested in. We will create a filter expression and place it between two flat brackets following the DataFrame we wish to filter.
 
</div>

In [43]:
accident_list[accident_list.state == my_state]

event_id  aircraft_id                    ntsb_make ntsb_model  \
66   20130102X35708            1              BELL HELICOPTER        407   
117  20060705X00868            1                         BELL       206B   
144  20140908X10448            1  ROBINSON HELICOPTER COMPANY     R44 II   

    ntsb_number  year               date        city state country  \
66   CEN13FA122  2013  01/02/13 00:00:00  CLEAR LAKE    IA     USA   
117  CHI06FA173  2006  06/30/06 00:00:00     WALFORD    IA     USA   
144  CEN14LA487  2014  09/06/14 00:00:00   MACEDONIA    IA     USA   

     total_fatalities  in_usa  pilot_total latimes_make latimes_model  count  \
66                3.0    True          1.0         BELL           407    3.0   
117               1.0    True          1.0         BELL           206  112.0   
144               1.0    True          1.0     ROBINSON           R44   22.0   

    latimes_make_and_model  
66                BELL 407  
117               BELL 206  
144           ROBINSON R44

<div style="max-width: 640px;">

Now we should save the results of that filter into a new variable separate from the full list we imported from the CSV file. Since it includes only the sites for the location we’re interested in let’s call it `my_accidents`.

</div>

In [46]:
my_accidents = accident_list[accident_list.state == my_state]

<div style="max-width: 640px;">

To check our work and find out how many committees are left after the filter, let’s run the DataFrame inspection commands we learned earlier.

First `head`.

</div>

In [47]:
my_accidents.head()

event_id  aircraft_id                    ntsb_make ntsb_model  \
66   20130102X35708            1              BELL HELICOPTER        407   
117  20060705X00868            1                         BELL       206B   
144  20140908X10448            1  ROBINSON HELICOPTER COMPANY     R44 II   

    ntsb_number  year               date        city state country  \
66   CEN13FA122  2013  01/02/13 00:00:00  CLEAR LAKE    IA     USA   
117  CHI06FA173  2006  06/30/06 00:00:00     WALFORD    IA     USA   
144  CEN14LA487  2014  09/06/14 00:00:00   MACEDONIA    IA     USA   

     total_fatalities  in_usa  pilot_total latimes_make latimes_model  count  \
66                3.0    True          1.0         BELL           407    3.0   
117               1.0    True          1.0         BELL           206  112.0   
144               1.0    True          1.0     ROBINSON           R44   22.0   

    latimes_make_and_model  
66                BELL 407  
117               BELL 206  
144           ROBINSON R44

Then `info`.

In [48]:
my_accidents.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3 entries, 66 to 144
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   event_id                3 non-null      object 
 1   aircraft_id             3 non-null      int64  
 2   ntsb_make               3 non-null      object 
 3   ntsb_model              3 non-null      object 
 4   ntsb_number             3 non-null      object 
 5   year                    3 non-null      int64  
 6   date                    3 non-null      object 
 7   city                    3 non-null      object 
 8   state                   3 non-null      object 
 9   country                 3 non-null      object 
 10  total_fatalities        3 non-null      float64
 11  in_usa                  3 non-null      bool   
 12  pilot_total             3 non-null      float64
 13  latimes_make            3 non-null      object 
 14  latimes_model           3 non-null      obj

## Merging two dataframes together

<div style="max-width: 640px;">

Next we'll cover how to merge two DataFrames together into a combined table. Before we can do that, we need to read in a second file. We'll pull `faa-survey.csv`, which contains annual estimates of how many hours each type of helicopter was in the air. It was acquired via a Freedom of Information Act request with the FAA.

We can rip it in the same was the NTSB accident list, with `read_csv`.
    
</div>

In [49]:
survey = pd.read_csv("faa-survey.csv")

<div style="max-width: 640px;">

When joining two tables together, the first step is to look carefully at the columns in each table to find a common column that can be joined. We can do that with the `info` command we learned earlier.

</div>

In [51]:
accident_list.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 163 entries, 0 to 162
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   event_id                163 non-null    object 
 1   aircraft_id             163 non-null    int64  
 2   ntsb_make               163 non-null    object 
 3   ntsb_model              163 non-null    object 
 4   ntsb_number             163 non-null    object 
 5   year                    163 non-null    int64  
 6   date                    163 non-null    object 
 7   city                    163 non-null    object 
 8   state                   162 non-null    object 
 9   country                 163 non-null    object 
 10  total_fatalities        163 non-null    float64
 11  in_usa                  163 non-null    bool   
 12  pilot_total             163 non-null    float64
 13  latimes_make            163 non-null    object 
 14  latimes_model           163 non-null    ob

In [52]:
survey.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   year                    120 non-null    int64  
 1   latimes_make            120 non-null    object 
 2   latimes_model           120 non-null    object 
 3   population_count        120 non-null    int64  
 4   active_count            120 non-null    float64
 5   total_hours             120 non-null    int64  
 6   latimes_make_and_model  120 non-null    object 
dtypes: float64(1), int64(3), object(3)
memory usage: 6.7+ KB


<div style="max-width: 640px;">

You can see that each table contains the `latimes_make_and_model` column. We can therefore join the two files using the pandas <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge.html">`merge`</a> method.

<div class="alert alert-block alert-warning"><p>If you are familar with traditional databases, you may recognize that the merge method in pandas is similar to SQL’s JOIN statement. If you dig into <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge.html">merge’s documentation</a> you will see it has many of the same options.</p></div>

Merging two DataFrames is as simple as passing both to pandas built-in `merge` method and specifying which field we’d like to use to connect them together. We will save the result into another new variable, which I'm going to call `merged_list`.
    
</div>

In [53]:
merged_list = pd.merge(accident_list, survey, on="latimes_make_and_model")

That new DataFrame can be inspected like any other.

In [54]:
merged_list.head()

event_id  aircraft_id ntsb_make ntsb_model ntsb_number  year_x  \
0  20121010X41638            1      BELL       206B  CEN13FA009    2012   
1  20121010X41638            1      BELL       206B  CEN13FA009    2012   
2  20121010X41638            1      BELL       206B  CEN13FA009    2012   
3  20121010X41638            1      BELL       206B  CEN13FA009    2012   
4  20121010X41638            1      BELL       206B  CEN13FA009    2012   

                date     city state country  ...  latimes_make_x  \
0  10/10/12 00:00:00  CROWLEY    LA     USA  ...            BELL   
1  10/10/12 00:00:00  CROWLEY    LA     USA  ...            BELL   
2  10/10/12 00:00:00  CROWLEY    LA     USA  ...            BELL   
3  10/10/12 00:00:00  CROWLEY    LA     USA  ...            BELL   
4  10/10/12 00:00:00  CROWLEY    LA     USA  ...            BELL   

   latimes_model_x  count latimes_make_and_model year_y  latimes_make_y  \
0              206  112.0               BELL 206   2006            BELL   
1              206  112.0               BELL 206   2007            BELL   
2              206  112.0               BELL 206   2008            BELL   
3              206  112.0               BELL 206   2009            BELL   
4              206  112.0               BELL 206   2010            BELL   

  latimes_model_y  population_count active_count total_hours  
0             206              1701       1549.0      652307  
1             206              1650       1448.0      589158  
2             206              1420       1303.0      518892  
3             206              1455       1337.0      477082  
4             206              1529       1436.0      579440  

[5 rows x 23 columns]

By looking at the columns you can check how many rows survived the merge.

In [38]:
merged_list.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 43211 entries, 0 to 43210
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   handle      43211 non-null  object
 1   name        43211 non-null  object
 2   url_x       43211 non-null  object
 3   location    43211 non-null  object
 4   timezone    43211 non-null  object
 5   identifier  43211 non-null  object
 6   file_name   43211 non-null  object
 7   url_y       43211 non-null  object
 8   mtime       43211 non-null  object
 9   size        43211 non-null  int64 
 10  md5         43211 non-null  object
 11  sha1        43211 non-null  object
dtypes: int64(1), object(11)
memory usage: 4.3+ MB


You can also see that the DataFrame now contains the same number of records as the `screenshot_list`, as well as all of the columns in both tables. Columns with the same name have had a suffix of `_x` or `_y` automatically appended to indicate whether they came from the first or second DataFrame submitted to the merge.

## Sorting

Another simple but common technique for analyzing data is sorting. This can be useful for ranking the DataFrame to show the highest and lowest members of the group according to a particular column. The [`sort_values`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html) is how pandas does it.

In [43]:
merged_list.sort_values("size")

handle                 name                         url_x  \
19552        laist                LAist            https://laist.com/   
39067    ukrinform            Ukrinform     https://www.ukrinform.ua/   
25567      News3LV          KSNV News 3          https://news3lv.com/   
23173     mercnews         Mercury News  https://www.mercurynews.com/   
18996     KyivPost             KyivPost     https://www.kyivpost.com/   
...            ...                  ...                           ...   
42519          wsj  Wall Street Journal          https://www.wsj.com/   
10451  fastcompany         Fast Company  https://www.fastcompany.com/   
10448  fastcompany         Fast Company  https://www.fastcompany.com/   
10449  fastcompany         Fast Company  https://www.fastcompany.com/   
10455  fastcompany         Fast Company  https://www.fastcompany.com/   

          location             timezone        identifier  \
19552  Los Angeles  America/Los_Angeles        laist-2022   
39067         Kiev          Europe/Kiev    ukrinform-2022   
25567    Las Vegas  America/Los_Angeles      news3lv-2022   
23173     San Jose  America/Los_Angeles     mercnews-2022   
18996         Kiev          Europe/Kiev     kyivpost-2022   
...            ...                  ...               ...   
42519     New York     America/New_York          wsj-2022   
10451     New York     America/New_York  fastcompany-2022   
10448     New York     America/New_York  fastcompany-2022   
10449     New York     America/New_York  fastcompany-2022   
10455     New York     America/New_York  fastcompany-2022   

                                              file_name  \
19552        laist-2022-06-24T18:37:43.658736-07:00.jpg   
39067    ukrinform-2022-03-24T15:38:42.807929+02:00.jpg   
25567      news3lv-2022-07-08T19:08:35.222676-07:00.jpg   
23173     mercnews-2022-05-02T06:45:41.641250-07:00.jpg   
18996     kyivpost-2022-04-21T16:42:48.004499+03:00.jpg   
...                                                 ...   
42519          wsj-2022-05-17T08:07:02.756004-04:00.jpg   
10451  fastcompany-2022-07-19T22:16:19.926420-04:00.jpg   
10448  fastcompany-2022-07-18T13:44:59.403155-04:00.jpg   
10449  fastcompany-2022-07-18T22:22:49.761458-04:00.jpg   
10455  fastcompany-2022-07-21T22:19:44.329803-04:00.jpg   

                                                   url_y                mtime  \
19552  https://archive.org/download/laist-2022/laist-...  2022-06-25 01:37:45   
39067  https://archive.org/download/ukrinform-2022/uk...  2022-03-24 13:38:43   
25567  https://archive.org/download/news3lv-2022/news...  2022-07-09 02:08:37   
23173  https://archive.org/download/mercnews-2022/mer...  2022-05-02 13:45:43   
18996  https://archive.org/download/kyivpost-2022/kyi...  2022-04-21 13:42:49   
...                                                  ...                  ...   
42519  https://archive.org/download/wsj-2022/wsj-2022...  2022-05-17 12:07:07   
10451  https://archive.org/download/fastcompany-2022/...  2022-07-20 02:16:21   
10448  https://archive.org/download/fastcompany-2022/...  2022-07-18 17:45:02   
10449  https://archive.org/download/fastcompany-2022/...  2022-07-19 02:22:50   
10455  https://archive.org/download/fastcompany-2022/...  2022-07-22 02:19:46   

         size                               md5  \
19552   13138  ea987d5a23910acc5fc3afce720baed4   
39067   14755  74383537eb3be7d4007c86ac3d6402a7   
25567   16092  0594aafe3e1d3df37d5ff4ca86821be3   
23173   16104  fb6547e22027b350c61b8f3ff961ea71   
18996   18029  c17661c54118426112255a15f2fe111a   
...       ...                               ...   
42519  569445  533fc96be915658c123ade2c2707f4f7   
10451  584409  9c32beefe1fd8b2ca109dd42ec3614ef   
10448  680645  84d2132c066f50aeadda10f71edcdd33   
10449  682426  44f890d84a0d92ff4ae28993dde0e627   
10455  687226  b4d4f21df49c026b9c3a0aa7d0f48844   

                                           sha1  
19552  c82e951aeb3af7b182407e2bb14a05efe6928b4d  
39067  

Note that returns the DataFrame resorted in ascending order from lowest to highest. That is pandas default way of sorting. Here's how you reverse it to show the largest values first.

In [44]:
merged_list.sort_values("size", ascending=False)

handle                 name                         url_x  \
10455  fastcompany         Fast Company  https://www.fastcompany.com/   
10449  fastcompany         Fast Company  https://www.fastcompany.com/   
10448  fastcompany         Fast Company  https://www.fastcompany.com/   
10451  fastcompany         Fast Company  https://www.fastcompany.com/   
42519          wsj  Wall Street Journal          https://www.wsj.com/   
...            ...                  ...                           ...   
18996     KyivPost             KyivPost     https://www.kyivpost.com/   
23173     mercnews         Mercury News  https://www.mercurynews.com/   
25567      News3LV          KSNV News 3          https://news3lv.com/   
39067    ukrinform            Ukrinform     https://www.ukrinform.ua/   
19552        laist                LAist            https://laist.com/   

          location             timezone        identifier  \
10455     New York     America/New_York  fastcompany-2022   
10449     New York     America/New_York  fastcompany-2022   
10448     New York     America/New_York  fastcompany-2022   
10451     New York     America/New_York  fastcompany-2022   
42519     New York     America/New_York          wsj-2022   
...            ...                  ...               ...   
18996         Kiev          Europe/Kiev     kyivpost-2022   
23173     San Jose  America/Los_Angeles     mercnews-2022   
25567    Las Vegas  America/Los_Angeles      news3lv-2022   
39067         Kiev          Europe/Kiev    ukrinform-2022   
19552  Los Angeles  America/Los_Angeles        laist-2022   

                                              file_name  \
10455  fastcompany-2022-07-21T22:19:44.329803-04:00.jpg   
10449  fastcompany-2022-07-18T22:22:49.761458-04:00.jpg   
10448  fastcompany-2022-07-18T13:44:59.403155-04:00.jpg   
10451  fastcompany-2022-07-19T22:16:19.926420-04:00.jpg   
42519          wsj-2022-05-17T08:07:02.756004-04:00.jpg   
...                                                 ...   
18996     kyivpost-2022-04-21T16:42:48.004499+03:00.jpg   
23173     mercnews-2022-05-02T06:45:41.641250-07:00.jpg   
25567      news3lv-2022-07-08T19:08:35.222676-07:00.jpg   
39067    ukrinform-2022-03-24T15:38:42.807929+02:00.jpg   
19552        laist-2022-06-24T18:37:43.658736-07:00.jpg   

                                                   url_y                mtime  \
10455  https://archive.org/download/fastcompany-2022/...  2022-07-22 02:19:46   
10449  https://archive.org/download/fastcompany-2022/...  2022-07-19 02:22:50   
10448  https://archive.org/download/fastcompany-2022/...  2022-07-18 17:45:02   
10451  https://archive.org/download/fastcompany-2022/...  2022-07-20 02:16:21   
42519  https://archive.org/download/wsj-2022/wsj-2022...  2022-05-17 12:07:07   
...                                                  ...                  ...   
18996  https://archive.org/download/kyivpost-2022/kyi...  2022-04-21 13:42:49   
23173  https://archive.org/download/mercnews-2022/mer...  2022-05-02 13:45:43   
25567  https://archive.org/download/news3lv-2022/news...  2022-07-09 02:08:37   
39067  https://archive.org/download/ukrinform-2022/uk...  2022-03-24 13:38:43   
19552  https://archive.org/download/laist-2022/laist-...  2022-06-25 01:37:45   

         size                               md5  \
10455  687226  b4d4f21df49c026b9c3a0aa7d0f48844   
10449  682426  44f890d84a0d92ff4ae28993dde0e627   
10448  680645  84d2132c066f50aeadda10f71edcdd33   
10451  584409  9c32beefe1fd8b2ca109dd42ec3614ef   
42519  569445  533fc96be915658c123ade2c2707f4f7   
...       ...                               ...   
18996   18029  c17661c54118426112255a15f2fe111a   
23173   16104  fb6547e22027b350c61b8f3ff961ea71   
25567   16092  0594aafe3e1d3df37d5ff4ca86821be3   
39067   14755  74383537eb3be7d4007c86ac3d6402a7   
19552   13138  ea987d5a23910acc5fc3afce720baed4   

                                           sha1  
10455  a13c74ecb785a5f09f08ac6803457afc1bf9c162  
10449  

You can limit the result to the top five by chaining the head method at the end.

In [45]:
merged_list.sort_values("size", ascending=False).head()

handle                 name                         url_x  \
10455  fastcompany         Fast Company  https://www.fastcompany.com/   
10449  fastcompany         Fast Company  https://www.fastcompany.com/   
10448  fastcompany         Fast Company  https://www.fastcompany.com/   
10451  fastcompany         Fast Company  https://www.fastcompany.com/   
42519          wsj  Wall Street Journal          https://www.wsj.com/   

       location          timezone        identifier  \
10455  New York  America/New_York  fastcompany-2022   
10449  New York  America/New_York  fastcompany-2022   
10448  New York  America/New_York  fastcompany-2022   
10451  New York  America/New_York  fastcompany-2022   
42519  New York  America/New_York          wsj-2022   

                                              file_name  \
10455  fastcompany-2022-07-21T22:19:44.329803-04:00.jpg   
10449  fastcompany-2022-07-18T22:22:49.761458-04:00.jpg   
10448  fastcompany-2022-07-18T13:44:59.403155-04:00.jpg   
10451  fastcompany-2022-07-19T22:16:19.926420-04:00.jpg   
42519          wsj-2022-05-17T08:07:02.756004-04:00.jpg   

                                                   url_y                mtime  \
10455  https://archive.org/download/fastcompany-2022/...  2022-07-22 02:19:46   
10449  https://archive.org/download/fastcompany-2022/...  2022-07-19 02:22:50   
10448  https://archive.org/download/fastcompany-2022/...  2022-07-18 17:45:02   
10451  https://archive.org/download/fastcompany-2022/...  2022-07-20 02:16:21   
42519  https://archive.org/download/wsj-2022/wsj-2022...  2022-05-17 12:07:07   

         size                               md5  \
10455  687226  b4d4f21df49c026b9c3a0aa7d0f48844   
10449  682426  44f890d84a0d92ff4ae28993dde0e627   
10448  680645  84d2132c066f50aeadda10f71edcdd33   
10451  584409  9c32beefe1fd8b2ca109dd42ec3614ef   
42519  569445  533fc96be915658c123ade2c2707f4f7   

                                           sha1  
10455  a13c74ecb785a5f09f08ac6803457afc1bf9c162  
10449  463a6c808e34785bca0a082f332cf0aced853340  
10448  7ac5a6a7293cc5323114ae219dc94e68e3952038  
10451  7af1fe7285af0cdcfd59a258b51ded54d1f49fd3  
42519  2f3d3012c0d3c7b4106c88cb2f046c58ad0b7919

## Grouping

The [`groupby`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html) method allows you to group a DataFrame by a column and then calculate a sum, or any other statistic, for each unique value. This functions much like the ["pivot table"](https://en.wikipedia.org/wiki/Pivot_table) feature found in most spreadsheets.

Let's use it to analyze the timezone field, which show where the organizations indexed by News Homepages are based. You start by passing its name to the function.

In [46]:
merged_list.groupby("timezone")

A nice start but you’ll notice you don’t get much back. The data’s been grouped, but we haven’t chosen what to do with it yet. If we wanted the file size total by zone, we would sum the `size` field the same way we did earlier for the entire DataFrame.

In [47]:
merged_list.groupby("timezone")['size'].sum()

timezone
America/Bogota            20359830
America/Buenos_Aires      24953179
America/Chicago         2284188701
America/Denver           329645851
America/Edmonton          19357058
America/Los_Angeles     2402403425
America/Mexico_City      350211916
America/Montevideo         7379875
America/New_York        4664240116
America/Phoenix           58965447
America/Vancouver         37972418
Asia/Beirut               25311438
Asia/Qatar                21678804
Asia/Tbilisi              28888169
Asia/Tokyo               102516510
Europe/Berlin              3371896
Europe/Dublin             18099916
Europe/Kiev              353249376
Europe/London            809543441
Europe/Madrid             28401302
Europe/Moscow             61413650
Europe/Oslo                9954086
Europe/Paris             580587143
Europe/Riga               76488230
US/Eastern               215213124
US/Hawaii                  4988770
Name: size, dtype: int64

Again our data has come back as an ugly Series. To reformat it as a pretty DataFrame use the `reset_index` method again.

In [48]:
merged_list.groupby("timezone")['size'].sum().reset_index()

timezone        size
0         America/Bogota    20359830
1   America/Buenos_Aires    24953179
2        America/Chicago  2284188701
3         America/Denver   329645851
4       America/Edmonton    19357058
5    America/Los_Angeles  2402403425
6    America/Mexico_City   350211916
7     America/Montevideo     7379875
8       America/New_York  4664240116
9        America/Phoenix    58965447
10     America/Vancouver    37972418
11           Asia/Beirut    25311438
12            Asia/Qatar    21678804
13          Asia/Tbilisi    28888169
14            Asia/Tokyo   102516510
15         Europe/Berlin     3371896
16         Europe/Dublin    18099916
17           Europe/Kiev   353249376
18         Europe/London   809543441
19         Europe/Madrid    28401302
20         Europe/Moscow    61413650
21           Europe/Oslo     9954086
22          Europe/Paris   580587143
23           Europe/Riga    76488230
24            US/Eastern   215213124
25             US/Hawaii     4988770

Next re-sort totals from highest to lowest. Remember the `sort_values` trick we learned earlier? That’ll do it.

In [49]:
merged_list.groupby("timezone")['size'].sum().reset_index().sort_values("size", ascending=False)

timezone        size
8       America/New_York  4664240116
5    America/Los_Angeles  2402403425
2        America/Chicago  2284188701
18         Europe/London   809543441
22          Europe/Paris   580587143
17           Europe/Kiev   353249376
6    America/Mexico_City   350211916
3         America/Denver   329645851
24            US/Eastern   215213124
14            Asia/Tokyo   102516510
23           Europe/Riga    76488230
20         Europe/Moscow    61413650
9        America/Phoenix    58965447
10     America/Vancouver    37972418
13          Asia/Tbilisi    28888169
19         Europe/Madrid    28401302
11           Asia/Beirut    25311438
1   America/Buenos_Aires    24953179
12            Asia/Qatar    21678804
0         America/Bogota    20359830
4       America/Edmonton    19357058
16         Europe/Dublin    18099916
21           Europe/Oslo     9954086
7     America/Montevideo     7379875
25             US/Hawaii     4988770
15         Europe/Berlin     3371896

## Computing

Here's how you can create a new column based on the data in other columns, a process sometimes known as “computing.”

Let's say you want to take an extra step beyond the last chapter and figure out the screenshot totals by month. One approach would be to create a column and fill it with the month of each screenshot. Then to group and count on that field.

One problem. before we can extract the month from the `mtime` field we need to convert it into a Python datetime object. Run `info` again and you'll see that pandas read it in as a string, which pandas prefers to call an `object`.

In [50]:
merged_list.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 43211 entries, 0 to 43210
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   handle      43211 non-null  object
 1   name        43211 non-null  object
 2   url_x       43211 non-null  object
 3   location    43211 non-null  object
 4   timezone    43211 non-null  object
 5   identifier  43211 non-null  object
 6   file_name   43211 non-null  object
 7   url_y       43211 non-null  object
 8   mtime       43211 non-null  object
 9   size        43211 non-null  int64 
 10  md5         43211 non-null  object
 11  sha1        43211 non-null  object
dtypes: int64(1), object(11)
memory usage: 4.3+ MB


Thankfully, pandas comes with a utility that is highly skilled at translating strings into timestamp. It's called `pd.to_datetime` and it expects a Series as an input. Using it to convert mtime looks something like this:

In [51]:
pd.to_datetime(merged_list.mtime)

0       2022-07-09 03:55:23
1       2022-07-09 15:38:13
2       2022-07-10 04:02:11
3       2022-07-10 15:38:47
4       2022-07-11 04:03:34
                ...        
43206   2022-07-20 04:25:48
43207   2022-07-20 16:51:07
43208   2022-07-21 04:34:37
43209   2022-07-21 17:01:36
43210   2022-07-22 04:38:18
Name: mtime, Length: 43211, dtype: datetime64[ns]

The result of that new column can be attached to the DataFrame by assigning to it a new column. We could call it anything. Let's go with `datetime`.

In [52]:
merged_list['datetime'] = pd.to_datetime(merged_list.mtime)

If you run `info` again you'll see the new column and the preferred data type.

In [53]:
merged_list.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 43211 entries, 0 to 43210
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   handle      43211 non-null  object        
 1   name        43211 non-null  object        
 2   url_x       43211 non-null  object        
 3   location    43211 non-null  object        
 4   timezone    43211 non-null  object        
 5   identifier  43211 non-null  object        
 6   file_name   43211 non-null  object        
 7   url_y       43211 non-null  object        
 8   mtime       43211 non-null  object        
 9   size        43211 non-null  int64         
 10  md5         43211 non-null  object        
 11  sha1        43211 non-null  object        
 12  datetime    43211 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(11)
memory usage: 4.6+ MB


Now that Python is treating the column as a datetime object, we can extract the month from each record in the series by accessing the `dt` attribute put there by pandas.

In [54]:
merged_list.datetime.dt.month

0        7
1        7
2        7
3        7
4        7
        ..
43206    7
43207    7
43208    7
43209    7
43210    7
Name: datetime, Length: 43211, dtype: int64

Those values can be saved into their own new computed column.

In [55]:
merged_list['month'] = merged_list.datetime.dt.month

Which will not be found by `info`.

In [56]:
merged_list.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 43211 entries, 0 to 43210
Data columns (total 14 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   handle      43211 non-null  object        
 1   name        43211 non-null  object        
 2   url_x       43211 non-null  object        
 3   location    43211 non-null  object        
 4   timezone    43211 non-null  object        
 5   identifier  43211 non-null  object        
 6   file_name   43211 non-null  object        
 7   url_y       43211 non-null  object        
 8   mtime       43211 non-null  object        
 9   size        43211 non-null  int64         
 10  md5         43211 non-null  object        
 11  sha1        43211 non-null  object        
 12  datetime    43211 non-null  datetime64[ns]
 13  month       43211 non-null  int64         
dtypes: datetime64[ns](1), int64(2), object(11)
memory usage: 4.9+ MB


Finally, summing the number of screenshots by month will require going back to `groupby`. We will introduce the `size` method, which can count the number of records in each group. Again, `reset_index` tranlates the result into a DataFrame.

In [57]:
merged_list.groupby("month").size().reset_index()

month      0
0      3   1080
1      4   4510
2      5   4623
3      6  12567
4      7  20431

## Further reading

With that, we've covered the basic skills necessary to access and analyze data with pandas. Or at least everything we could fit in two quick hours. If you'd like to learn more, consult [the full edition of "First Python Notebook"](https://palewi.re/docs/first-python-notebook/).